Increase minimum debug.log size to 10MB after shrink.

morcos opened the pull request:
`-shrinkdebugfile` defaults to true if no `-debug` options are passed, but 200k is an insufficient history to help users who may have restarted their nodes debug any issues.

morcos committed:
Increase minimum debug.log size to 10MB after shrink.

MarcoFalke commented:
utACK 338ad5e0a53bc153454bf1f3dc46de0f705419d2

On Tue, Jan 24, 2017 at 6:50 PM, Alex Morcos <notifications@github.com> wrote:
> -shrinkdebugfile defaults to true if no -debug options are passed, but 200k
> is an insufficient history to help users who may have restarted their nodes
> debug any issues.
>
> ________________________________
>
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/bitcoin/bitcoin/pull/9625
>
> Commit Summary
>
> Increase minimum debug.log size to 9MB after shrink.
>
> File Changes
>
> M src/util.cpp (3)
>
> Patch Links:
>
> https://github.com/bitcoin/bitcoin/pull/9625.patch
> https://github.com/bitcoin/bitcoin/pull/9625.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

paveljanik commented:
utACK https://github.com/bitcoin/bitcoin/pull/9625/commits/338ad5e0a53bc153454bf1f3dc46de0f705419d2

btcdrak commented:
utACK 338ad5e

gmaxwell commented:
Yea, sure, though this also won't be enough in many cases, I think. utACK

sipa commented:
Concept ACK, but can you introduce a named constant for this?

jonasschnelli commented:
utACK 338ad5e0a53bc153454bf1f3dc46de0f705419d2, agree about the const.

MarcoFalke commented:
We might even extend the command line option to specify how many mb
the user wants to store, but better leave the nits for past-0.15
branch-off.

TheBlueMatt commented:
Looks good. Would also be nice to add a comment above where this is called in init noting - "Do this first since it both loads a bunch of debug.log into memory, and because this needs to happen before any other debug.log printing".

morcos commented:
I think I addressed the feedback.

Incidentally I was looking into file_size and fseek to see whether I needed to have the 10% buffer, and it didn't look to me from the internets that the negative fseek from SEEK_END was necessarily a supported operation?

MarcoFalke commented on src/util.cpp:
```diff
@@ -723,13 +723,17 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
-    if (file && boost::filesystem::file_size(pathLog) > 10 * 1000000)
+    // If debug.log file is more than 10% bigger the RECENT_DEBUG_HISTORY_SIZE
+    // trim it down by saving only the last RECENT_DEBUG_HISTORY_SIZE bytes
+    if (file && boost::filesystem::file_size(pathLog) > 11 * (RECENT_DEBUG_HISTORY_SIZE / 10))
```
RECENT_DEBUG_HISTORY_SIZE * 11 / 10 might give better results for the integer division, when replaced by a command line option. Not that it matters, but µnit...

MarcoFalke commented on src/util.cpp:
```diff
@@ -723,13 +723,17 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
     boost::filesystem::path pathLog = GetDataDir() / "debug.log";
     FILE* file = fopen(pathLog.string().c_str(), "r");
-    if (file && boost::filesystem::file_size(pathLog) > 10 * 1000000)
+    // If debug.log file is more than 10% bigger the RECENT_DEBUG_HISTORY_SIZE
```
s/the/than/, maybe? Or even better, don't mention the name of the constant and just write. `If debug.log exceeds the target size by 10% or more, trim it to the target size.` or similar. (It is not like the code here is hard to understand and needs excessive developer documentation)

re-ACK otherwise.

laanwj commented:
> Incidentally I was looking into file_size and fseek to see whether I needed to have the 10% buffer, and it didn't look to me from the internets that the negative fseek from SEEK_END was necessarily a supported operation?

I'm 100% sure it is. I can't find it explicitly in the man page, but it is the only thing `SEEK_END` is useful for, you'd hardly ever want to use it with a positive offset. Generally it is used for file formats that have a directory structure at the end such as ZIP.

laanwj commented:
This seemed ready for merge. @MarcoFalke's last comments seem to be small nits that could be done as a separate patch if necessary.
